{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.10","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"import numpy as np\nimport tensorflow as tf\nimport keras\nfrom keras.models import Sequential\nfrom keras.layers import Conv2D,MaxPooling2D,Dense,Flatten,Dropout\nimport matplotlib.pyplot as plt\n\nfrom tensorflow.keras.layers import BatchNormalization\nfrom keras_preprocessing import image\nfrom tensorflow.keras.preprocessing.image import ImageDataGenerator\n\n\nimport os\nimport cv2","metadata":{"_uuid":"d41d8301-2d7c-41ab-9cdb-e99ff60c66fd","_cell_guid":"b1eb822e-89bd-42d4-853a-f14948f28cc1","collapsed":false,"jupyter":{"outputs_hidden":false},"execution":{"iopub.status.busy":"2023-05-24T14:54:41.905041Z","iopub.execute_input":"2023-05-24T14:54:41.905681Z","iopub.status.idle":"2023-05-24T14:57:12.022835Z","shell.execute_reply.started":"2023-05-24T14:54:41.905633Z","shell.execute_reply":"2023-05-24T14:57:12.020080Z"},"trusted":true},"execution_count":3,"outputs":[{"name":"stdout","text":"\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a64c0344bd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/mtcnn/\u001b[0m\n\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a64c0412310>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/mtcnn/\u001b[0m\n\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a64c04126d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/mtcnn/\u001b[0m\n\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a64c0412a90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/mtcnn/\u001b[0m\n\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a64c0412e50>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/mtcnn/\u001b[0m\n\u001b[31mERROR: Could not find a version that satisfies the requirement mtcnn\u001b[0m\n\u001b[31mERROR: No matching distribution found for mtcnn\u001b[0m\n","output_type":"stream"},{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-3-a435a68cc000>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'pip install mtcnn'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mmtcnn\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mMTCNN\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcv2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'mtcnn'"],"ename":"ModuleNotFoundError","evalue":"No module named 'mtcnn'","output_type":"error"}]},{"cell_type":"code","source":"train_dir=\"../input/face-recognition-dataset/Original Images/Original Images/\"\ngenerator = ImageDataGenerator()\ntrain_ds = generator.flow_from_directory(train_dir,target_size=(224, 224),batch_size=32)\nclasses = list(train_ds.class_indices.keys())","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.024310Z","iopub.status.idle":"2023-05-24T14:57:12.024827Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"model = Sequential()\nmodel.add(Conv2D(32, kernel_size = (3, 3), activation='relu', input_shape=(224,224,3)))\nmodel.add(MaxPooling2D(pool_size=(2,2)))\nmodel.add(BatchNormalization())\nmodel.add(Conv2D(64, kernel_size=(3,3), activation='relu'))\nmodel.add(MaxPooling2D(pool_size=(2,2)))\nmodel.add(BatchNormalization())\nmodel.add(Conv2D(64, kernel_size=(3,3), activation='relu'))\nmodel.add(MaxPooling2D(pool_size=(2,2)))\nmodel.add(BatchNormalization())\nmodel.add(Conv2D(96, kernel_size=(3,3), activation='relu'))\nmodel.add(MaxPooling2D(pool_size=(2,2)))\nmodel.add(BatchNormalization())\nmodel.add(Conv2D(32, kernel_size=(3,3), activation='relu'))\nmodel.add(MaxPooling2D(pool_size=(2,2)))\nmodel.add(BatchNormalization())\nmodel.add(Dropout(0.2))\nmodel.add(Flatten())\nmodel.add(Dense(128, activation='relu'))\n#model.add(Dropout(0.3))\nmodel.add(Dense(len(classes),activation='softmax'))","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.028936Z","iopub.status.idle":"2023-05-24T14:57:12.029406Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"model.compile(\n    loss = 'categorical_crossentropy',\n    optimizer = 'adam',\n    metrics = [\"accuracy\"])\nmodel.summary()","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.030323Z","iopub.status.idle":"2023-05-24T14:57:12.030796Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"history = model.fit(train_ds,epochs= 15, batch_size=32)","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.031883Z","iopub.status.idle":"2023-05-24T14:57:12.032355Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"plt.plot(history.history['accuracy'])\nplt.plot(history.history['loss'])\nplt.xlabel('Time')\nplt.legend(['accuracy', 'loss'])\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.033423Z","iopub.status.idle":"2023-05-24T14:57:12.033906Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"def predict_image(image_path):\n    img = image.load_img(image_path, target_size=(224,224,3))\n    plt.imshow(img)\n    plt.show()\n    x = image.img_to_array(img)\n    x = np.expand_dims(x, axis=0)\n    images = np.vstack([x])\n    pred = model.predict(images, batch_size=32)\n    print(\"Actual: \"+(image_path.split(\"/\")[-1]).split(\"_\")[0])\n    print(\"Predicted: \"+classes[np.argmax(pred)])","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.035164Z","iopub.status.idle":"2023-05-24T14:57:12.035630Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Brad Pitt/Brad Pitt_102.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.036516Z","iopub.status.idle":"2023-05-24T14:57:12.037007Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Charlize Theron/Charlize Theron_26.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.038074Z","iopub.status.idle":"2023-05-24T14:57:12.038543Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Henry Cavill/Henry Cavill_28.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.039602Z","iopub.status.idle":"2023-05-24T14:57:12.040113Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Tom Cruise/Tom Cruise_27.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.041138Z","iopub.status.idle":"2023-05-24T14:57:12.041596Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Robert Downey Jr/Robert Downey Jr_106.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.042533Z","iopub.status.idle":"2023-05-24T14:57:12.043018Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Natalie Portman/Natalie Portman_25.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.044111Z","iopub.status.idle":"2023-05-24T14:57:12.044573Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Lisa Kudrow/Lisa Kudrow_34.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.045598Z","iopub.status.idle":"2023-05-24T14:57:12.046082Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Ellen Degeneres/Ellen Degeneres_20.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.046998Z","iopub.status.idle":"2023-05-24T14:57:12.047510Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Dwayne Johnson/Dwayne Johnson_29.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.048548Z","iopub.status.idle":"2023-05-24T14:57:12.049060Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"predict_image(\"../input/face-recognition-dataset/Original Images/Original Images/Elizabeth Olsen/Elizabeth Olsen_11.jpg\")","metadata":{"execution":{"iopub.status.busy":"2023-05-24T14:57:12.050054Z","iopub.status.idle":"2023-05-24T14:57:12.050778Z"},"trusted":true},"execution_count":null,"outputs":[]}]}